File: java/L000/LeetCode0081.java
package L000;

public class LeetCode0081 {
    public boolean search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        for (int l = 0, r = nums.length - 1;l <= r;) {
            int m = l + (r - l) / 2;
            if (nums[m] == target) {
                return true;
            }
            if (nums[m] < nums[r]) {
                if (target > nums[m] && target <= nums[r]) {
                    l = m + 1;
                } else {
                    r = m - 1;
                }
            } else if (nums[m] > nums[l]) {
                if (target < nums[m] && target >= nums[l]) {
                    r = m - 1;
                } else {
                    l = m + 1;
                }
            } else if (nums[m] == nums[l] && nums[m] == nums[r]) {
                for (int i = l;i <= r;++i) {
                    if (nums[i] == target) {
                        return true;
                    }
                }
                return false;
            } else if (nums[m] == nums[l]) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
        return false;
    }
}

File: cpp/L000/LeetCode0081.cc
#include <vector>
using std::vector;

class Solution {
public:
    bool search(vector<int>& nums, int target) {
        for (int low = 0, high = nums.size() - 1;low <= high;) {
            int mid = low + (high - low) / 2;
            if (nums[mid] == target) {
                return true;
            }
            if (nums[low] == nums[high] && nums[low] == nums[mid]) {
                ++low;
                --high;
            } else if (nums[low] <= nums[mid]) {
                if (nums[low] <= target && target <= nums[mid]) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            } else if (nums[mid] <= target && target <= nums[high]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }      
        return false;
    }
};

int main() {
    Solution s;
    vector<int> r{1,0,1,1,1};
    s.search(r, 0);
    return 0;
}